raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansProcessor.java

```diff
@@ -196,7 +242,101 @@ public void process(Record<TraceIdentity, RawSpan> record) {
         .record(Duration.between(start, Instant.now()).toMillis(), TimeUnit.MILLISECONDS);
     // no need to do context.forward. the punctuator will emit the trace once it's eligible to be
     // emitted
-    return;
+  }
+
+  private void processSpanForPeerServiceNameIdentification(
+      TraceIdentity key, RawSpan value, TraceState traceState, long currentTimeMs) {
+    Event event = value.getEvent();
+    String tenantId = key.getTenantId();
+    ByteBuffer traceId = value.getTraceId();
+    boolean firstEntry = (traceState == null);
+    Optional<String> maybeEnvironment = HttpSemanticConventionUtils.getEnvironmentForSpan(event);
+    if (SpanSemanticConventionUtils.isClientSpanForOCFormat(
+        event.getAttributes().getAttributeMap())) {
+      handleClientSpan(tenantId, event, maybeEnvironment.orElse(null));
+    } else {
+      handleServerSpan(tenantId, event, maybeEnvironment.orElse(null));
+    }
+
+    Timestamps timestamps =
+        traceLatencyMeter.trackEndToEndLatencyTimestamps(
+            currentTimeMs, firstEntry ? currentTimeMs : traceState.getTraceStartTimestamp());
+    StructuredTrace trace =
+        StructuredTraceBuilder.buildStructuredTraceFromRawSpans(
+            List.of(value), traceId, tenantId, timestamps);
+    context.forward(new Record<>(key, trace, currentTimeMs), OUTPUT_TOPIC_PRODUCER);
+  }
+
+  // put the peer service identity and corresponding service name in state store
+  private void handleClientSpan(String tenantId, Event event, String environment) {
+    Optional<String> maybeHostAddr = HttpSemanticConventionUtils.getHostIpAddress(event);
+    Optional<String> maybePeerAddr = HttpSemanticConventionUtils.getPeerIpAddress(event);
+    Optional<String> maybePeerPort = HttpSemanticConventionUtils.getPeerPort(event);
+    String serviceName = event.getServiceName();
+    PeerIdentity peerIdentity =
+        PeerIdentity.newBuilder()
+            .setIpIdentity(
+                IpIdentity.newBuilder()
+                    .setTenantId(tenantId)
+                    .setEnvironment(environment)
+                    .setHostAddr(maybeHostAddr.orElse(null))
+                    .setPeerAddr(maybePeerAddr.orElse(null))
+                    .setPeerPort(maybePeerPort.orElse(null))
+                    .build())
+            .build();
+    if (PeerIdentityValidator.isValid(peerIdentity)) {
+      this.peerIdentityToSpanMetadataStateStore.put(
+          peerIdentity,
+          SpanMetadata.newBuilder()
+              .setServiceName(serviceName)
+              .setEventId(HexUtils.getHex(event.getEventId()))
```

Comment:
Any specific reason to use `hex` form here? We can just store eventId, right?

Reply:
No, will update it